src/thor/triplegbuilder.cc

```diff
@@ -1042,6 +946,328 @@ void AccumulateRecostingInfoForward(const valhalla::Options& options,
+   * @param graphtile   Graph tile of the current edge.
+   * @param controller  Controller specifying attributes to add to trip edge.
+   *
+   */
+  void AddTransitNodes(TripLeg_Node* trip_node,
+                       const NodeInfo* node,
+                       const GraphId& startnode,
+                       const GraphTile* start_tile,
+                       const GraphTile* graphtile,
+                       const AttributesController& controller) {
+
+    if (node->type() == NodeType::kTransitStation) {
+      const TransitStop* transit_station =
+          start_tile->GetTransitStop(start_tile->node(startnode)->stop_index());
+      TransitStationInfo* transit_station_info = trip_node->mutable_transit_station_info();
+
+      if (transit_station) {
+        // Set onstop_id if requested
+        if (controller.attributes.at(kNodeTransitStationInfoOnestopId) &&
+            transit_station->one_stop_offset()) {
+          transit_station_info->set_onestop_id(
+              graphtile->GetName(transit_station->one_stop_offset()));
+        }
+
+        // Set name if requested
+        if (controller.attributes.at(kNodeTransitStationInfoName) && transit_station->name_offset()) {
+          transit_station_info->set_name(graphtile->GetName(transit_station->name_offset()));
+        }
+
+        // Set latitude and longitude
+        LatLng* stop_ll = transit_station_info->mutable_ll();
+        // Set transit stop lat/lon if requested
+        if (controller.attributes.at(kNodeTransitStationInfoLatLon)) {
+          PointLL ll = node->latlng(start_tile->header()->base_ll());
+          stop_ll->set_lat(ll.lat());
+          stop_ll->set_lng(ll.lng());
+        }
+      }
+    }
+
+    if (node->type() == NodeType::kTransitEgress) {
+      const TransitStop* transit_egress =
+          start_tile->GetTransitStop(start_tile->node(startnode)->stop_index());
+      TransitEgressInfo* transit_egress_info = trip_node->mutable_transit_egress_info();
+
+      if (transit_egress) {
+        // Set onstop_id if requested
+        if (controller.attributes.at(kNodeTransitEgressInfoOnestopId) &&
+            transit_egress->one_stop_offset()) {
+          transit_egress_info->set_onestop_id(graphtile->GetName(transit_egress->one_stop_offset()));
+        }
+
+        // Set name if requested
+        if (controller.attributes.at(kNodeTransitEgressInfoName) && transit_egress->name_offset()) {
+          transit_egress_info->set_name(graphtile->GetName(transit_egress->name_offset()));
+        }
+
+        // Set latitude and longitude
+        LatLng* stop_ll = transit_egress_info->mutable_ll();
+        // Set transit stop lat/lon if requested
+        if (controller.attributes.at(kNodeTransitEgressInfoLatLon)) {
+          PointLL ll = node->latlng(start_tile->header()->base_ll());
+          stop_ll->set_lat(ll.lat());
+          stop_ll->set_lng(ll.lng());
+        }
+      }
+    }
+  }
+
+  void AddTransitInfo(TripLeg_Node* trip_node,
```

Comment:
nit: wonder if we need struct here? Maybe adding `AddTransitInfo` method next to the other existed ones is enough
Or is it used to group all three

Reply:
yep that was my first thought but that method specifically keeps state in the main loop of triplegbuilder::build. so i took that state and moved it into the struct to encapsulate it